doc, test: test and explain service flag handling

mzumsande opened the pull request:
Service flags received from the peer-to-peer network are handled differently, depending on how we receive them.
If received directly from an outbound peer the flags belong to, they replace existing flags.
If received via gossip relay (so that anyone could send them), new flags are added, but existing ones but cannot be overwritten.

Document that and add test coverage for it.

mzumsande committed:
doc, test: Test and explain service flag handling

Service flags are handled differently, depending on whether
validated (if received from the peer) or unvalidated (received
via gossip relay).

brunoerg commented:
Concept ACK

jonatack reviewed (commented):
ACK d536e5a6325d1885224f36debdcc5245b94efe8a

mzumsande commented on src/addrman.h:
```diff
@@ -111,6 +111,9 @@ class AddrMan
 
     /**
      * Attempt to add one or more addresses to addrman's new table.
+     * If an address already exists in addrman, the existing entry may be updated
+     * (e.g. adding additional service flags). If the existing entry is in the new table,
```
I think that followed from the `@return` doc, but it wasn't so obvious that adding an existing addr to another new bucket also affects the return value; Changed the `@return` description to `true if at least one address is successfully added, or added to an additional bucket. Unaffected by updates.`

mzumsande commented on src/net_processing.cpp:
```diff
@@ -3376,6 +3376,9 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         {
+            // Overwrites potentially existing services. In contrast to this,
+            // unvalidated services received via gossip relay are only
+            // ever added but cannot replace existing ones.
```
done

mzumsande commented on src/addrman.h:
```diff
@@ -111,6 +111,9 @@ class AddrMan
      * Attempt to add one or more addresses to addrman's new table.
+     * If an address already exists in addrman, the existing entry may be updated
+     * (e.g. adding additional service flags). If the existing entry is in the new table,
+     * it may be added to more buckets, improving the probability of selection.
```
I think it's better if the header doesn't refer to implementation details. If I remember correctly, this logic used to be in `Add()`, then `Add_()`, then `AddSingle()`, and I don't think it's nice if we have to update `addrman.h` for refactors like that.

mzumsande commented on src/test/addrman_tests.cpp:
```diff
@@ -1095,6 +1095,32 @@ BOOST_AUTO_TEST_CASE(addrman_update_address)
+    BOOST_CHECK_EQUAL(vAddr5.size(), 1U);
+    BOOST_CHECK_EQUAL(vAddr5.at(0).nServices, NODE_NETWORK);
+
+    // Adding service flags even works when the addr is in Tried
```
similar to above, I don't like referring to implementation details in comments unless really necessary.

mzumsande commented:
[d536e5a ](https://github.com/bitcoin/bitcoin/commit/d536e5a6325d1885224f36debdcc5245b94efe8a)to [74ebd4d](https://github.com/bitcoin/bitcoin/commit/74ebd4d1359edce82a134dfcd3da9840f8d206e2): Addressed feedback

furszy reviewed (commented):
ACK 74ebd4d1359edce82a134dfcd3da9840f8d206e2

brunoerg reviewed (approved):
utACK 74ebd4d1359edce82a134dfcd3da9840f8d206e2

achow101 commented:
ACK 74ebd4d1359edce82a134dfcd3da9840f8d206e2
